src/components/LiquidityMining/Create/Steps/PairAndReward/index.tsx

```diff
@@ -1,101 +1,223 @@
-  reward: TokenAmount | null
-  onLiquidityPairChange: (liquidityPair: Pair) => void
-  onRewardTokenChange: (token: Token) => void
+interface TokenAndLimitProps {
+  stakeTokenOrPair: Pair | Token | undefined | null
+  unlimitedPool: boolean
+  campaingType: CampaignType
+  setStakeTokenOrPair: (liquidityPair: Pair | Token | null) => void
+  onStakingCapChange: (newValue: TokenAmount | null) => void
+  onUnlimitedPoolChange: (newValue: boolean) => void
 }
 
-export default function PairAndReward({
-  liquidityPair,
-  reward,
-  onLiquidityPairChange,
-  onRewardTokenChange,
-}: PairAndRewardProps) {
+export default function StakeTokenAndLimit({
+  stakeTokenOrPair,
+  unlimitedPool,
+  setStakeTokenOrPair,
+  campaingType,
+  onStakingCapChange,
+  onUnlimitedPoolChange,
+}: TokenAndLimitProps) {
   const [pairSearchOpen, setPairSearchOpen] = useState<boolean>(false)
   const [currencySearchOpen, setCurrencySearchOpen] = useState<boolean>(false)
+  const [stakingCapString, setStakingCapString] = useState('')
 
-  const handleOpenPairSearch = useCallback(() => {
-    setPairSearchOpen(true)
+  const inputRef = React.useRef<HTMLInputElement>(null)
+
+  useEffect(() => {
+    if (unlimitedPool) {
+      setStakingCapString('')
+      onStakingCapChange(null)
+    }
+  }, [onStakingCapChange, stakeTokenOrPair, unlimitedPool])
+
+  const handelOpenPairOrTokenSearch = useCallback(value => {
+    if (value === CampaignType.PAIR) {
+      setPairSearchOpen(true)
+    } else {
+      setCurrencySearchOpen(true)
+    }
   }, [])
 
-  const handleDismissPairSearch = useCallback(() => {
-    setPairSearchOpen(false)
+  const handleDismissTokenOrPairSelection = useCallback((type: CampaignType) => {
+    if (type === CampaignType.TOKEN) setCurrencySearchOpen(false)
+    else setPairSearchOpen(false)
   }, [])
 
   const handlePairSelection = useCallback(
     selectedPair => {
-      onLiquidityPairChange(selectedPair)
+      setStakeTokenOrPair(selectedPair)
     },
-    [onLiquidityPairChange]
+    [setStakeTokenOrPair]
   )
 
-  const handleOpenCurrencySearch = useCallback(() => {
-    setCurrencySearchOpen(true)
-  }, [])
-
-  const handleDismissCurrencySearch = useCallback(() => {
-    setCurrencySearchOpen(false)
-  }, [])
+  useEffect(() => {
+    setStakeTokenOrPair(null)
+  }, [campaingType, setStakeTokenOrPair])
 
-  const handleCurrencySelection = useCallback(
-    selectedCurrency => {
-      onRewardTokenChange(selectedCurrency)
+  const widthValue = useMemo(() => {
+    if (stakingCapString.length > 0 && inputRef.current) return inputRef.current.clientWidth
+    else return 0
+  }, [stakingCapString, inputRef])
+  const handleLocalStakingCapChange = useCallback(
+    rawValue => {
+      if (!stakeTokenOrPair || (stakeTokenOrPair instanceof Pair && !stakeTokenOrPair.liquidityToken)) return
+      setStakingCapString(rawValue)
+      const tokenOrPair = stakeTokenOrPair instanceof Token ? stakeTokenOrPair : stakeTokenOrPair.liquidityToken
```

Comment:
We are using `stakeTokenOrPair instanceof X` to many times. Can you check it once and use a variable to check if `isPair` or `isToken`?

Reply:
yeah I kinda have that issue across the board I will try to debug it but simply making a variable that checks for type doesn't address it since it will just return boolean value and not actually take what type it is into account